Add a Makefile that can be used to drive EMake for simple builds

vermiculus opened the pull request:
Fix #17

vermiculus committed:
Add a Makefile that can be used to drive EMake for simple builds

vermiculus commented:
@DamienCassou Let me know your thoughts – does this Makefile do enough to suit your needs? I reviewed a lot of my packages and it would work on those, so if it would also work for you, I'll know it's not just my style that it works with 😄

DamienCassou commented:
This looks great. Good job! Some feedback after trying on libmpdel:

1. `test-buttercup` fails if there is no buttercup test
1. files `-autoloads.el` and `-pkg.el` should be ignored. I suggest this change:

    ```diff
    @@ -9,7 +9,8 @@ endif
     #   EMACS_ARGS: [extra arguments for each invocation of emacs]
     #   (below; see README)
     PACKAGE_FILE		?= $(PACKAGE_BASENAME).el
    -PACKAGE_LISP		?= $(wildcard $(PACKAGE_BASENAME)*.el)
    +PACKAGE_LISP		?= $(filter-out %-pkg.el %-autoloads.el, $(wildcard $(PACKAGE_BASENAME)*.el))
    +
     PACKAGE_TESTS		?= $(wildcard test/*.el)
     PACKAGE_ARCHIVES	?= gnu melpa
     PACKAGE_TEST_ARCHIVES	?= gnu melpa
    ```

1. I think CI scripts should ignore dir-local variables:

    ```diff
    @@ -21,7 +22,7 @@ EMAKE = PACKAGE_FILE="$(PACKAGE_FILE)" \
     	PACKAGE_ARCHIVES="$(PACKAGE_ARCHIVES)" \
     	PACKAGE_TEST_DEPS="$(PACKAGE_TEST_DEPS)" \
     	PACKAGE_TEST_ARCHIVES="$(PACKAGE_TEST_ARCHIVES)" \
    -	emacs -batch -l emake.el $(EMACS_ARGS) --eval "(emake (pop argv))"
    +	emacs -batch -l emake.el $(EMACS_ARGS) --eval "(setq enable-dir-local-variables nil)" --eval "(emake (pop argv))"
     
     # Set up our phony targets so Make doesn't think there are files by
     # these names.
    ```

1. it seems that checkdoc and package-lint are not executed on test files. Is that desired?
1. can you please add compilation with error-on-warn to the `test` rule?

DamienCassou commented:
You might also want to add `emake.el` as a dependency to the `.elpa/` rule.

vermiculus commented:
It would've been a little easier if you had written line comments 😅 but I'll try to tackle them one at a time.

1. I've removed the `test` target. It's a convenience target anyway, and one that isn't really hard to just do by hand in `.travis.yml`:
    ```yaml
    script:
      - make -f emake.mk lint-checkdoc
      - make -f emake.mk lint-package-lint
      - make -f emake.mk test-ert
      - make -f emake.mk test-buttercup
    ```
    or you could include `emake.mk` in your own Makefile and setup dependencies:
    ```makefile
    include emake.mk
    emake.mk:
    	wget ...
    test: lint-checkdoc lint-package-lint test-buttercup
    ```
2. I've added this, thanks!
3. Normally would agree, but others may not. I've added it before the expansion of `EMACS_ARGS`
4. I do not think test files should be under the same standards of the actual package files (at least in this sphere, where most products are pet projects). If this is desired, it can easily be added with two or three lines in a Makefile that create a new target for this purpose:
    ```makefile
    include emake.mk
    emake.mk:
    	wget …
    lint-tests: PACKAGE_LISP := $(PACKAGE_TESTS)
    lint-tests: lint-checkdoc lint-package-lint
    test: lint-checkdoc lint-package-lint lint-tests test-ert
    ```
5. Nope, compilation is not testing. It can be added as above, but I don't think it's appropriate to do this in the standard Makefile.

vermiculus committed:
Use emake.mk in our own Makefile

DamienCassou commented on emake.mk:
```diff
@@ -0,0 +1,101 @@
+PACKAGE_TEST_ARCHIVES   ?= gnu
+
+# Then, make it easy to invoke Emacs with EMake loaded.
+EMAKE = PACKAGE_FILE="$(PACKAGE_FILE)" \
```
what about adding a `@` before `PACKAGE_FILE` to avoid getting all the noise in the log?

vermiculus replied:
I can understand the desire to keep a clean log, but hiding information causes more problems than it solves in my experience. If we were to do this, I would want EMake to report its invocation pattern under the `emake:` prefix.

Actually I'm not sure I'd want to do even that – from the documentation of variable `command-line-args`, it doesn't look I can rely on it to hold the entire invocation.

DamienCassou replied:
ok, I lost this battle :-).

DamienCassou commented on emake.mk:
```diff
@@ -0,0 +1,101 @@
+# these names.
+.PHONY: clean setup install compile test help
+
+help:                           ## show help
```
this doesn't seem to work for me

```
make help
emake.mk                       show help
emake.mk                       clean all generated files
emake.mk                       clean all generated files including scripts
emake.mk                       Commands useful for Travis
...
```

vermiculus replied:
Yeah I noticed that… I need to dig into how the pipeline is actually behaving. It works when there's only one Makefile in the list, but it seems to break here.

vermiculus replied:
Fixed! Added the `-h` option to the grep call.

DamienCassou replied:
thanks

DamienCassou commented on emake.mk:
```diff
@@ -0,0 +1,101 @@
+# available.
+setup: emacs emake.el
+
+install: .elpa/                 ## install dependencies
```
I'm not sure about the name of this rule. Usually, `make install` installs the project under a common directory such as /usr/local. What about `install-dependencies`, `setup-dependencies`, `setup-elpa`...?

vermiculus replied:
I'm wondering now if a 'target pseudonym' like this is warranted…

Is `make .elpa/` clear?

All the testing targets already require `.elpa/`.

DamienCassou replied:
to me `make .elpa` is clear enough.

DamienCassou commented on emake.mk:
```diff
@@ -0,0 +1,101 @@
+## Support targets
+
+emake.el:                       ## download the EMake script
+	wget 'https://raw.githubusercontent.com/vermiculus/emake.el/master/emake.el'
```
if it's ok, could you get `curl` instead of `wget` here? My CI environment doesn't have `wget` by default but I've always seen `curl` everywhere.

vermiculus replied:
Absolutely – what would be the appropriate curl invocation?

vermiculus replied:
nvm, I just copied your invocation from mpdel 😄

DamienCassou replied:
sorry, I wanted to write it on the previous message but forgot :-(. This is what [magit uses](https://github.com/magit/magit/blob/188dec9c47e5e63bb52f405f0b81c78693161a28/.travis.yml#L5) except that I replaced short options by long ones for clarity.

```
CURL = curl --fail --silent --show-error --insecure --location --retry 9 --retry-delay 9

emake.el:
	$(CURL) -O https://raw.githubusercontent.com/vermiculus/emake.el/batteries-included/emake.el
```

vermiculus replied:
Updated, thanks 😄

DamienCassou commented:
I consider the content of the PR good enough to migrate both libmpdel and mpdel for which I created 2 PRs:

- https://gitlab.petton.fr/mpdel/libmpdel/merge_requests/25
- https://gitlab.petton.fr/mpdel/mpdel/merge_requests/28
